pkg/controller/controller.go

```diff
@@ -0,0 +1,238 @@
+package controller
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"sync/atomic"
+
+	machinev1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	certificatesv1 "k8s.io/api/certificates/v1beta1"
+	certificatesv1beta1 "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	configNamespace    = "openshift-config-managed"
+	kubeletCAConfigMap = "csr-controller-ca"
+)
+
+// MachineApproverReconciler reconciles a machine-approver  object
+type CertificateApprover struct {
+	client.Client
+	RestCfg *rest.Config
+	Config  ClusterMachineApproverConfig
+}
+
+func (m *CertificateApprover) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return m.buildWithManager(mgr, options, m)
+}
+
+func (m *CertificateApprover) buildWithManager(mgr ctrl.Manager, options controller.Options, c reconcile.Reconciler) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(options).
+		For(&certificatesv1.CertificateSigningRequest{}, builder.WithPredicates(predicate.Funcs{
+			CreateFunc:  func(e event.CreateEvent) bool { return pendingCertFilter(e.Object) },
+			UpdateFunc:  func(e event.UpdateEvent) bool { return pendingCertFilter(e.ObjectNew) },
+			GenericFunc: func(e event.GenericEvent) bool { return pendingCertFilter(e.Object) },
+			DeleteFunc:  func(e event.DeleteEvent) bool { return false },
+		})).
+		Watches(
+			&source.Kind{Type: &corev1.ConfigMap{}},
+			handler.EnqueueRequestsFromMapFunc(m.toCSRs),
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc:  func(e event.CreateEvent) bool { return caConfigMapFilter(e.Object, nil) },
+				UpdateFunc:  func(e event.UpdateEvent) bool { return caConfigMapFilter(e.ObjectOld, e.ObjectNew) },
+				GenericFunc: func(e event.GenericEvent) bool { return pendingCertFilter(e.Object) },
```

Comment:
Should this one be `caConfigMapFilter` too? Potential copy/paste error

Reply:
Fixed, thanks @JoelSpeed